### PR TITLE
fix unresolved inline functions with newer compiler versions

### PR DESCRIPTION
**Pull request by stevstrong2**
completes #677

### CODE REVIEW

**Comment by stevstrong2** [comment]
With these changes both compilers (4.8.3 and 7.2.1) build warning and error free.

**Comment by dewhisna** [comment]
Well, this change works for me now on PlatformIO, though some of the `warning: always_inline function might not be inlinable [-Wattributes]` warnings you were trying to fix in the #677 PR came back.  Didn't you mean to make the `__attribute__((always_inline))` entries in this commit `static inline` instead?

**Comment by dewhisna** [comment]
BTW, the lack of the Arduino IDE properly rebuilding things when testing stuff like this is one of the reasons I abandoned it long ago....

**Comment by stevstrong** [comment]
Now I really cannot reproduce the warnings.
Do you use 7.2.1? Windows or Linux?
I want to avoid static for inline functions because then they will get space allocated in flash for each module in which they are included.

**Comment by dewhisna** [comment]
Your argument to "avoid using static for inline functions" doesn't make any sense and especially not for ones with the attribute to tell the compiler to force it to be inlined, because 'inline' means put a whole separate copy of the entire function everywhere it's being called -- not just once in every module included, but every single call to the function (i.e. it "inlines" the content of the function in place of the function call).  `static` by itself just allocates one per module, and `static inline` means to `inline` it (i.e. copy per every call of the function) when possible and if not possible, have one per module (i.e. library being linked), but not escape the module or library boundary.

BTW, I'm running Linux though it really shouldn't matter if you have all warnings enabled and did a clean build.  Here's the warnings I get after applying the changes from your new PR:

```
Compiling .pio/build/genericSTM32F103VG/FrameworkArduino/libmaple/exti.c.o
In file included from ~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/cores/maple/libmaple/dma_f1.c:40:0:
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/dma_private.h:40:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dma_irq_handler(dma_dev *dev, dma_tube tube)
                                     ^~~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/cores/maple/libmaple/exti.c:225:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_extis(uint32 start, uint32 stop) {
                                     ^~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/cores/maple/libmaple/exti.c:211:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_single_exti(uint32 exti) {
                                     ^~~~~~~~~~~~~~~~~~~~
```

```
Compiling .pio/build/genericSTM32F103VG/FrameworkArduino/libmaple/usart_f1.c.o
In file included from ~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/cores/maple/libmaple/timer.c:36:0:
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:185:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_general(timer_dev *dev) {
                                     ^~~~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:171:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_adv_cc(timer_dev *dev) {
                                     ^~~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:156:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_adv_trg_com(timer_dev *dev) {
                                     ^~~~~~~~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:152:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_adv_up(timer_dev *dev) {
                                     ^~~~~~~~~~~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:148:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_adv_brk(timer_dev *dev) {
                                     ^~~~~~~~~~~Compiling .pio/build/genericSTM32F103VG/FrameworkArduino/libmaple/usart_private.c.o
~~~~~
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/timer_private.h:124:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void dispatch_single_irq(timer_dev *dev,
                                     ^~~~~~~~~~~~~~~~~~~
In file included from ~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/cores/maple/libmaple/usart_f1.c:37:0:
~/.platformio/packages/framework-arduinoststm32-custom/STM32F1/system/libmaple/usart_private.h:40:37: warning: always_inline function might not be inlinable [-Wattributes]
 __attribute__((always_inline)) void usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs) {
                                     ^~~~~~~~~
```

**Comment by stevstrong** [comment]
> 'inline' means put a whole separate copy of the entire function everywhere it's being called -- not just once in every module included, but every single call to the function (i.e. it "inlines" the content of the function in place of the function call)

Exactly this is the goal. They are functions which should be inlined within the ISRs.

By applying my fix, have you also applied the #677? If not then you should.
So you better get a fresh copy of it and then apply this, too.

**Comment by dewhisna** [comment]
Yes, I also applied #677.  In fact, I was on the branch where I had previously merged #677 when testing, and then merged this PR on top of it.

Since you explicitly want it always inlined, I think you are just missing the keyword to actually tell it to inline as you only have the attribute and no `inline` keyword, see [https://stackoverflow.com/questions/32432596/warning-always-inline-function-might-not-be-inlinable-wattributes](https://stackoverflow.com/questions/32432596/warning-always-inline-function-might-not-be-inlinable-wattributes)

In other words, I think you want to do this:
```diff
-__attribute__((always_inline)) void dispatch_single_exti(uint32 exti) {
+__attribute__((always_inline)) inline void dispatch_single_exti(uint32 exti) {
```

... for each place you have the attribute -- that is, add the `inline` keyword.  Otherwise, the `always_inline` attribute is confused because your function isn't actually declared as `inline`.  In other words, the `__attribute__((always_inline))` doesn't imply `inline`.

I think the compiler does that so that it's portable, as other compilers don't use the same attribute values and you'd otherwise be missing the `inline` keyword.

File: STM32F1/cores/maple/libmaple/exti.c [hunk 1]
```diff
@@ -208,6 +206,46 @@ void exti_do_select(__IO uint32 *exti_cr, exti_num num, exti_cfg port) {
     *exti_cr = cr;
 }
 
+/* This dispatch routine is for non-multiplexed EXTI lines only; i.e.,
+ * it doesn't check EXTI_PR. */
+__attribute__((always_inline)) void dispatch_single_exti(uint32 exti) {
```
Comment (dewhisna) [inline STM32F1/cores/maple/libmaple/exti.c]:
```diff
-__attribute__((always_inline)) void dispatch_single_exti(uint32 exti) {
+__attribute__((always_inline)) inline void dispatch_single_exti(uint32 exti) {
```

File: STM32F1/cores/maple/libmaple/exti.c [hunk 2]
```diff
@@ -208,6 +206,46 @@ void exti_do_select(__IO uint32 *exti_cr, exti_num num, exti_cfg port) {
...
+    }
+
+    handler(exti_channels[exti].arg);
+    EXTI_BASE->PR = (1U << exti);
+    asm volatile("nop");
+    asm volatile("nop");
+}
+
+/* Dispatch routine for EXTIs which share an IRQ. */
+__attribute__((always_inline)) void dispatch_extis(uint32 start, uint32 stop) {
```
Comment (dewhisna) [inline STM32F1/cores/maple/libmaple/exti.c]:
```diff
-__attribute__((always_inline)) void dispatch_extis(uint32 start, uint32 stop) {
+__attribute__((always_inline)) inline void dispatch_extis(uint32 start, uint32 stop) {
```

File: STM32F1/system/libmaple/dma_private.h [hunk 3]
```diff
@@ -37,7 +37,7 @@
 /* Wrap this in an ifdef to shut up GCC. (We provide DMA_GET_HANDLER
  * in the series support files, which need dma_irq_handler().) */
 #ifdef DMA_GET_HANDLER
-inline void dma_irq_handler(dma_dev *dev, dma_tube tube)
+__attribute__((always_inline)) void dma_irq_handler(dma_dev *dev, dma_tube tube)
```
Comment (dewhisna) [inline STM32F1/system/libmaple/dma_private.h]:
```diff
-__attribute__((always_inline)) void dma_irq_handler(dma_dev *dev, dma_tube tube)
+__attribute__((always_inline)) inline void dma_irq_handler(dma_dev *dev, dma_tube tube)
```

File: STM32F1/system/libmaple/timer_private.h [hunk 4]
```diff
@@ -121,7 +121,7 @@
  * line may be shared with another timer. For example, the timer 1
  * update interrupt shares an IRQ line with the timer 10 interrupt on
  * STM32F1 (XL-density), STM32F2, and STM32F4. */
-inline void dispatch_single_irq(timer_dev *dev,
+__attribute__((always_inline)) void dispatch_single_irq(timer_dev *dev,
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_single_irq(timer_dev *dev,
+__attribute__((always_inline)) inline void dispatch_single_irq(timer_dev *dev,
```

File: STM32F1/system/libmaple/timer_private.h [hunk 5]
```diff
@@ -145,15 +145,15 @@ inline void dispatch_single_irq(timer_dev *dev,
         }                                                               \
     } while (0)
 
-inline void dispatch_adv_brk(timer_dev *dev) {
+__attribute__((always_inline)) void dispatch_adv_brk(timer_dev *dev) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_adv_brk(timer_dev *dev) {
+__attribute__((always_inline)) inline void dispatch_adv_brk(timer_dev *dev) {
```

File: STM32F1/system/libmaple/timer_private.h [hunk 6]
```diff
@@ -145,15 +145,15 @@ inline void dispatch_single_irq(timer_dev *dev,
...
     dispatch_single_irq(dev, TIMER_BREAK_INTERRUPT, TIMER_SR_BIF);
 }
 
-inline void dispatch_adv_up(timer_dev *dev) {
+__attribute__((always_inline)) void dispatch_adv_up(timer_dev *dev) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_adv_up(timer_dev *dev) {
+__attribute__((always_inline)) inline void dispatch_adv_up(timer_dev *dev) {
```

File: STM32F1/system/libmaple/timer_private.h [hunk 7]
```diff
@@ -145,15 +145,15 @@ inline void dispatch_single_irq(timer_dev *dev,
...
     dispatch_single_irq(dev, TIMER_UPDATE_INTERRUPT, TIMER_SR_UIF);
 }
 
-inline void dispatch_adv_trg_com(timer_dev *dev) {
+__attribute__((always_inline)) void dispatch_adv_trg_com(timer_dev *dev) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_adv_trg_com(timer_dev *dev) {
+__attribute__((always_inline)) inline void dispatch_adv_trg_com(timer_dev *dev) {
```

File: STM32F1/system/libmaple/timer_private.h [hunk 8]
```diff
@@ -168,7 +168,7 @@ inline void dispatch_adv_trg_com(timer_dev *dev) {
     regs->SR = ~handled;
 }
 
-inline void dispatch_adv_cc(timer_dev *dev) {
+__attribute__((always_inline)) void dispatch_adv_cc(timer_dev *dev) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_adv_cc(timer_dev *dev) {
+__attribute__((always_inline)) inline void dispatch_adv_cc(timer_dev *dev) {
```

File: STM32F1/system/libmaple/timer_private.h [hunk 9]
```diff
@@ -182,7 +182,7 @@ inline void dispatch_adv_cc(timer_dev *dev) {
     regs->SR = ~handled;
 }
 
-inline void dispatch_general(timer_dev *dev) {
+__attribute__((always_inline)) void dispatch_general(timer_dev *dev) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/timer_private.h]:
```diff
-__attribute__((always_inline)) void dispatch_general(timer_dev *dev) {
+__attribute__((always_inline)) inline void dispatch_general(timer_dev *dev) {
```

File: STM32F1/system/libmaple/usart_private.h [hunk 10]
```diff
@@ -37,7 +37,7 @@
 #include <libmaple/ring_buffer.h>
 #include <libmaple/usart.h>
 
-inline void usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs) {
+__attribute__((always_inline)) void usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs) {
```
Comment (dewhisna) [inline STM32F1/system/libmaple/usart_private.h]:
```diff
-__attribute__((always_inline)) void usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs) {
+__attribute__((always_inline)) inline void usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs) {
```

**Comment by stevstrong** [comment]
Donna, have you checked that this does not introduce the same issue which was the origin for #551 ?

**Comment by dewhisna** [comment]
lol ... this is like playing whack-a-mole, only with compilers.  No, nor do I have a good way to verify all of the scenarios described in #551 . 
I tell you what -- with this pull as it stands, even though it generates a half-dozen warnings, does compile, link, and appears to run correctly with my gcc 7.2.1.  So I would be OK with the way you have it.  I can always work around spurious warnings and selectively disable them if need-be.  Others, with more limited build environments, may not be able to do that.  So go with whatever works best for the majority.

**Comment by stevstrong2** [comment]
I double checked the head with this PR again:
no errors and no warnings with Arduino IDE, GCC versions 4.8.3, 7.2.1, 7.3.1.

**Comment by stevstrong** [comment]
As we cannot support other IDE than Arduino IDE, and there are no errors and warnings with Arduino, I will merge this.

**Pull request merged**